source/common/ssl/context_config_impl.cc

```diff
@@ -45,32 +46,60 @@ Secret::TlsCertificateConfigProviderSharedPtr getTlsCertificateConfigProvider(
   return nullptr;
 }
 
+Secret::CertificateValidationContextConfigProviderSharedPtr
+getCertificateValidationContextConfigProviderFromSds(
+    Server::Configuration::TransportSocketFactoryContext& factory_context,
+    const envoy::api::v2::auth::SdsSecretConfig& sds_secret_config,
+    const envoy::api::v2::auth::CertificateValidationContext* default_cvc) {
+  if (!sds_secret_config.has_sds_config()) {
+    // static secret
+    auto secret_provider =
+        factory_context.secretManager().findStaticCertificateValidationContextProvider(
+            sds_secret_config.name());
+    if (!secret_provider) {
+      throw EnvoyException(fmt::format("Unknown static certificate validation context: {}",
+                                       sds_secret_config.name()));
+    }
+    return secret_provider;
+  } else {
+    auto secret_provider =
+        factory_context.secretManager().findOrCreateCertificateValidationContextProvider(
+            sds_secret_config.sds_config(), sds_secret_config.name(), factory_context);
+    if (default_cvc != nullptr) {
```

Comment:
Isn't this semantically wrong? If secretManager() find existing CertificateValidationContextProvider then you're overriding the existing default CVC. If you have multiple cluster with same sds config name but different default_cvc, (I think this is common in istio use case, root cert from SDS are same, but secure naming is different) the last one wins and applies to all cluster.

Reply:
That is a good catch. Thanks, Lizan! I modify findOrCreateCertificateValidationContextProvider() to add default cvc as extra key. If we have multiple Clusters which use different default cvc and same SDS config, then secret manager creates different secret providers. Only clusters with the same cvc and SDS config share one provider.

Reply:
OK it makes this correct in Envoy side, but in this case you're flooding SDS server with O(# of unique clusters + # of unique listeners) request with same fetch trusted CA CVC request. Do you want to optimize that a bit?

Reply:
@lizan I have updated this PR with a change to provider::secret() interface. Now we keep secret config (e.g. TlsCertificateConfigPtr and CertificateValidationContextConfigPtr into ContextConfigImpl) Secret provider only returns secret() in protobuf message. The secret config will be created when SdsApi::update_callback_manager_ is invoked. Could you please take a look? Thanks!

Reply:
@lizan I will add / update tests when this change looks good to proceed.